### PR TITLE
29 code review

### PR DESCRIPTION
**Pull request by Sean1572**
Intended for jay to do code review over. DO NOT ACTUALLY MERGE THIS!

### CODE REVIEW

File: create_dataset/create_dataset.py [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+        all_data = pd.read_csv(result_file, index_col=0)
+    else:
+        all_data = pd.DataFrame()
+    # walk dir to list paths to each original wav file
+    wav_file_paths = get_paths(wav_dir)
+    # open human label file
+    labels = pd.read_csv(labels)
+    use_2017 = None
+    # iterate through each individual original wav
+    if "2017" in labels['DATE'].iloc[0]:
```
Comment (jaradfor) [inline create_dataset/create_dataset.py:64]: it's a little unclear what this is checking for and what the downstream side effect is

File: create_dataset/create_dataset.py [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+        else:
+            filtered_labels = filter_labels_2018(wav,
+                                                 labels)
+        # output the labeled segments and return the dataframe of annotations
+        new_buow_rows = create_segments(wav,
+                                        filtered_labels,
+                                        output_dir,
+                                        class_list)
+        # create same number of noise segments from the same wav file randomly
+        all_buow_rows = create_noise_segments(wav,
```
Comment (jaradfor) [inline create_dataset/create_dataset.py:85]: since the data set is so strongly labelled with every five second interval having either a positive or negative label, shouldn't we extract all negative labels and leave it as a decision downstream to balance the classes?  Even if every batch is balanced, there is some benefit from having a larger sample pool of negative examples to draw from

File: create_dataset/create_dataset.py [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+    PARSER.add_argument('-labels', type=str,
+                        help='Path to human labeled csv')
+    PARSER.add_argument('-wav_dir', type=str,
+                        help='Path to directory containing wav files.')
+    PARSER.add_argument('-output_dir', type=str,
+                        help='Path to desired directory for segments.')
+    PARSER.add_argument('-class_list', type=str,
+                        help='Path to txt file of list of labeled classes')
+    ARGS = PARSER.parse_args()
+    main(ARGS.labels, ARGS.wav_dir, ARGS.output_dir, ARGS.class_list)
```
Comment (jaradfor) [inline create_dataset/create_dataset.py:134]: I don't know if this is the correct place for it, but since this file already has so much specific knowledge about the burrowing owl dataset, it might also be a good idea to include some sanity checks and basic tests here as well. If not here, it might be a good idea to put that somewhere nearby

File: create_dataset/create_segments.py [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+    wavs_file_paths = []
+    for path, _, files in os.walk(home_dir):
+        for file in files:
+            if file.endswith('.wav'):
+                new_file = os.path.join(path, file)
+                wavs_file_paths.append(new_file)
+    return wavs_file_paths
+
+
+def create_segments(wav, filtered_labels, out_path, class_list):
```
Comment (jaradfor) [inline create_dataset/create_segments.py:32]: is this function dataset specific? What are the assumptions placed on the input?

File: create_dataset/create_segments.py [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+        filtered_labels (pd.Dataframe): The human label file reduced
+            to only contain the rows of detections pertinent to the
+            wav of interest.
+        out_path (str): Path to directory where segment will be saved.
+        class_list (str): Path to the class list that you'd like segments
+            to be created for. What the manual ID's are in the human
+            label file- will ignore everything that is misspelled or
+            unknown labels.
+
+    Returns:
```
Comment (jaradfor) [inline create_dataset/create_segments.py:47]: What does this metadata look like?  Could be obvious  and answered elsewhere and I'm missing something

File: create_dataset/create_segments.py [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+        classes = next(reader)
+    print(classes)
+    try:
+        audio = AudioSegment.from_wav(wav)
+    except exceptions.CouldntDecodeError:
+        print(f"Couldn't decode: {wav}, moving to next file")
+    filtered_labels['MANUAL ID*'] = filtered_labels['MANUAL ID*'].str.lower()
+    print(filtered_labels)
+    df_row = 0
+    for _, row in filtered_labels.iterrows():
```
Comment (jaradfor) [inline create_dataset/create_segments.py:78]: a couple comments here would be nice but it's overall clear what is occurring

File: create_dataset/create_segments.py [hunk 7]
```diff
@@ -0,0 +1,167 @@
...
+    except exceptions.CouldntDecodeError:
+        print(f"Couldn't decode: {wav}, moving to next file")
+    call_type = "no_buow"
+    num = len(new_buow_rows) * 2
+    seconds_array = np.zeros(duration)
+    for _, row in new_buow_rows.iterrows():
+        start = int((row['segment_rel_start_ms'] / 1000) - 1)
+        end = int((row['segment_rel_start_ms'] / 1000)
+                  + row['segment_duration_s'])
+        mask_start = max(0, start - 30)
```
Comment (jaradfor) [inline create_dataset/create_segments.py:137]: is this the buffer that is mentioned? Some magic numbers, maybe this should be an input or constant, comments or better variable names clarifying this would also be nice

File: create_dataset/create_dataset.py (hunk 1 above)
Comment (jaradfor) [reply create_dataset/create_dataset.py:64]: if we were iterating over all of the wave files, this would need to be inside of the loop with my current understanding. but since this is outside of the loop, I'm guessing the 2017 and 2018  datasets are processed separately and then glued together somewhere else

File: whoot_model_training/README.md [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+1) Create a copy of the config found in `configs/config.yml` and fill it out for your dataset. See the [config](#config) section
+2) Edit train.py to set up training for your dataset. If you are using a new dataset which an extractor does not exist for, contact code authors. 
+3) run `python train.py path/to/your/config/file.yml`
+
+# Config
+
+## Default Config Properties
+The properties of `config.yml` are as follows:
+### Data paths
+`metadata_csv`: the path to the metadata file for your dataset.
```
Comment (jaradfor) [inline whoot_model_training/README.md:27]: maybe an inline example here would be good

File: whoot_model_training/README.md [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+
+`SUBPROJECT_NAME` is either "binary" or "multilabelClass"
+`DATASET_NAME` is buowset0
+
+# Repo Philosophy  
+
+The most challenging issue with machine learning is the dataset. This training repo intends to make it easy to modularize parts of the training pipeline, and integrate them together, ideally regardless of the dataset. 
+
+The pipeline works in 5 parts:
+- Extractors: Extractors take in raw data and reformats it into `AudioDatasets`, apache-arrow data structures implemented via HuggingFace with common columns between any dataset. Every label is one_hot_encoded and treated as multilabel regardless of the problem. Audio filepaths as casted into [Audio columns](https://huggingface.co/docs/datasets/v3.6.0/en/package_reference/main_classes#datasets.Audio). Extractors are *unique for each dataset* but *uniform in the AudioDataset*.
```
Comment (jaradfor) [inline whoot_model_training/README.md:49]: one hot encoded versus multi label, is this intentional or is multi class the correct description?

File: whoot_model_training/train.py [hunk 10]
```diff
@@ -0,0 +1,167 @@
...
+
+    # Run training
+    training_args = WhootTrainingArguments(
+        run_name=run_name,
+        subproject_name=config["SUBPROJECT_NAME"],
+        dataset_name=config["DATASET_NAME"],
+    )
+
+    # COMMON OPTIONAL ARGS
+    training_args.num_train_epochs = 2
```
Comment (jaradfor) [inline whoot_model_training/train.py:125]: It's great to have these but these should probably be put somewhere else, maybe default values?

File: whoot_model_training/train_binary.py [hunk 11]
```diff
@@ -0,0 +1,163 @@
+"""Trains a Mutliclass Model with Pytorch and Huggingface.
```
Comment (jaradfor) [inline whoot_model_training/train_binary.py:1]: What is the reason for this being separated from the other file train.py?

File: whoot_model_training/whoot_model_training/dataset.py [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+        Raises:
+            AssertionError if dataset is not correctly formatted.
+        """
+        for split in ds.keys():
+            dataset = ds[split]
+            for column in DEFAULT_COLUMNS:
+                phrase_one = "The column `"
+                phrase_two = "` is missing from dataset split `"
+                phrase_three = "`. Required by system"
+                state = (
```
Comment (jaradfor) [inline whoot_model_training/whoot_model_training/dataset.py:52]: not a real problem at all but this f string could be  made a lot more pythonic

File: whoot_model_training/train.py [hunk 13]
```diff
@@ -0,0 +1,167 @@
...
+    #         p=0.8
+    #     )
+    # ])
+
+    # Online preprocessors prepare data for training
+    train_preprocessor = MelModelInputPreprocessor(
+        TimmInputs, duration=3
+    )
+
+    preprocessor = MelModelInputPreprocessor(
```
Comment (jaradfor) [inline whoot_model_training/train.py:109]: needs better variable name

**Review by jaradfor — COMMENTED** [review]
Overall I think a lot of this code is very nice.  it should be very helpful when working with the burring owl and hopefully more datasets. I think the readmes could do with some more work explicitly tying things together, and detailing how the different parts of the script interact, as well as clarifying which files/functions are dataset specific and which ones are more general.  There are some minor magic numbers like one thousand across the code in a number of places.  If we are hoping to use this training script for  many different models across many different datasets, it might make a lot of sense to use a framework like pytorch lightning which could allow more freedom if it's ever needed. this current script looks completely sufficient for training any given model on a single dataset.  Using pytorch lightning might be a waste of effort though, if the hugging face trainer and the written extension really do handle the overwhelming majority of uses.  I think if you consider all of the datasets that we could want to use like birdset and custom ones and cant think of a  situation where  the given trainer would not work but pytorch lightning would, we're  probably ok. I don't know how the given code supports things like learning rate scheduling and other modern day training tricks.

 I think it would be a very good idea to add somewhere dataset specific testing, to make sure that the extractor written works. I also think some unit and integration tests for the datasets would be an extremely good idea, as dataset problems can be very sneaky and hard to resolve.

File: whoot_model_training/whoot_model_training/dataset.py (hunk 12 above)
Comment (Sean1572) [reply whoot_model_training/whoot_model_training/dataset.py:52]: I wish the linter complained about it

File: whoot_model_training/train.py (hunk 10 above)
Comment (Sean1572) [reply whoot_model_training/train.py:125]: Maybe we debated it. 

Train.py, in my mind, is intended to be a more flexible file. A user edits it, and that becomes the file that outlines an experiment (note this file gets saved to comet-ml every run).

File: whoot_model_training/README.md (hunk 9 above)
Comment (Sean1572) [reply whoot_model_training/README.md:49]: Where is the versus here? One-hot encoding applies to both multi-label and multi-class ([110] vs [100])

File: whoot_model_training/README.md (hunk 8 above)
Comment (Sean1572) [reply whoot_model_training/README.md:27]: True, I'll make an issue out of this

File: create_dataset/create_segments.py (hunk 7 above)
Comment (Sean1572) [reply create_dataset/create_segments.py:137]: @kgarwoodsdzwa

File: whoot_model_training/train_binary.py (hunk 11 above)
Comment (Sean1572) [reply whoot_model_training/train_binary.py:1]: See my explanation above for the intention of train.py

Basically, a different experiment, and an example to show how train.py can be used as an experiment config. 

Perhaps the metacommentary here is this intent is not well documented
